snapcraft/plugins/ruby.py

```diff
@@ -0,0 +1,160 @@
+        self._gems = options.gems or []
+        self._use_bundler = options.use_bundler or False
+        self._ruby_version_dir = None
+
+        self.build_packages.extend(['gcc', 'g++', 'make', 'zlib1g-dev',
+                                    'libssl-dev', 'libreadline-dev'])
+
+        version_map = [('2.%d.[0-9]' % i, '2.%d.0' % i) for i in range(5)]
+
+        for version_regex, version_dir in version_map:
+            if re.compile(version_regex).match(self._ruby_version):
+                self._ruby_version_dir = version_dir
+                break
+        # Raise exception when version_dir not found
+        if not self._ruby_version_dir:
+            raise SnapcraftEnvironmentError(
+                'Ruby version dir not in version_map.')
+
+    def snap_fileset(self):
+        fileset = super().snap_fileset()
+        fileset.append('-include/')
+        fileset.append('-share/')
+        return fileset
+
+    def pull(self):
+        super().pull()
+        os.makedirs(self._ruby_part_dir, exist_ok=True)
+        self._ruby_tar.download()
+        self._ruby_install(builddir=self._ruby_part_dir)
+        self._gem_install()
+        if self._use_bundler:
+            self._bundle_install()
+
+    def env(self, root):
+        env = super().env(root)
+        env.append('PATH={}:{}'.format(
+            os.path.join(root, 'bin'), os.environ['PATH']))
+        env.append('RUBYPATH={}'.format(os.path.join(root, 'bin')))
+        rubydir = os.path.join(root, 'lib', 'ruby')
+        rubylib = os.path.join(rubydir, self._ruby_version_dir)
+        env.append('RUBYLIB={}:{}'.format(
+            rubylib, os.path.join(
+                rubylib, '{}-linux'.format(platform.machine()))))
+        env.append('GEM_HOME={}'.format(
+            os.path.join(rubydir, 'gems', self._ruby_version_dir)))
+        env.append('GEM_PATH={}'.format(
+            os.path.join(rubydir, 'gems', self._ruby_version_dir)))
+        return env
+
+    def _ruby_install(self, builddir):
+        self._ruby_tar.provision(
+            builddir, clean_target=False, keep_tarball=True)
+        self.run(['./configure', '--disable-install-rdoc', '--prefix=/'],
+                 cwd=builddir)
+        self.run(['make', '-j{}'.format(self.parallel_build_count)],
+                 cwd=builddir)
+        self.run(['make', 'install', 'DESTDIR={}'.format(self.installdir)],
+                 cwd=builddir)
+        # Fix all shebangs to use the in-snap ruby
+        file_utils.replace_in_file(
+            self.installdir,
+            re.compile(r''),
+            re.compile(r'^#!.*ruby'),
+            r'#!/usr/bin/env ruby')
+
+    def _gem_install(self):
+        if self._use_bundler:
+            self._gems = self._gems + ['bundler']
+        if self._gems:
+            gem_install_cmd = [os.path.join(self.installdir, 'bin', 'ruby'),
+                               os.path.join(self.installdir, 'bin', 'gem'),
+                               'install']
```

Comment:
You'll need to add `--env-shebang` to the arglist here to make the gems usable (it adds the `/usr/bin/env` shebang instead of hard-coding a path to the part's installdir).